Implement enough of array slicing to call .len() on arrays

dario23 opened the pull request.

dario23 committed:
Implement enough of array slicing to call .len() on arrays

Aurel300 commented on prusti-interface/src/environment/mir_utils/split_aggregate_assignment.rs:
```diff
@@ -54,7 +54,9 @@ impl<'tcx> SplitAggregateAssignment<'tcx> for mir::Statement<'tcx> {
             mir::Rvalue::Use(_) |
-            mir::Rvalue::Ref(_, _, _) =>
+            mir::Rvalue::Ref(_, _, _) |
+            // slice creation is ok™
```
```suggestion
            // slice creation is ok
```

Aurel300 commented on prusti-interface/src/environment/polonius_info.rs:
```diff
@@ -593,7 +593,22 @@ fn get_borrowed_places<'a, 'tcx: 'a>(
+                                return Ok(match operand {
+                                    mir::Operand::Copy(ref place) |
+                                        mir::Operand::Move(ref place) => vec![place],
+                                        mir::Operand::Constant(_) => vec![],
```
```suggestion
                                    mir::Operand::Copy(ref place)
                                    | mir::Operand::Move(ref place) => vec![place],
                                    mir::Operand::Constant(_) => vec![],
```

dario23 committed:
Try to better specify which unsize casts to allow

dario23 committed:
hackily ignore unsupported raw_ref invariant

the invariants (that are always empty) are generated on the raw
variable, but other preconditions are generated on .val_ref, so we run
into consistency errors otherwise.

dario23 committed:
Return unsupported error for iterator methods early

They would encode to viper without crashing just fine, but then have
subtle other problems, so just avoid them for now.

dario23 committed:
Update testcase with iterators message

Aurel300 committed:
Update prusti-interface/src/environment/mir_utils/split_aggregate_assignment.rs
